core/store/src/trie/trie_storage.rs

```diff
@@ -13,25 +13,120 @@ use std::cell::RefCell;
 use std::io::ErrorKind;
 
 #[derive(Clone)]
-pub struct TrieCache(Arc<Mutex<LruCache<CryptoHash, Vec<u8>>>>);
+pub struct SyncTrieCache(Arc<Mutex<TrieCache>>);
+
+struct TrieCache {
+    cache_state: CacheState,
+    shard_cache: LruCache<CryptoHash, Vec<u8>>,
+    chunk_cache: HashMap<CryptoHash, Vec<u8>>,
+}
+
+enum CachePosition {
+    None,
+    ShardCache(Vec<u8>),
+    ChunkCache(Vec<u8>),
+}
+
+enum CacheState {
+    CachingShard,
+    CachingChunk,
+}
+
+pub enum RetrievalCost {
+    Free,
+    Full,
+}
 
 impl TrieCache {
+    fn get_cache_position(&mut self, hash: &CryptoHash) -> CachePosition {
+        match self.chunk_cache.get(hash) {
+            Some(value) => CachePosition::ChunkCache(value.clone()),
+            None => match self.shard_cache.get(hash) {
+                Some(value) => CachePosition::ShardCache(value.clone()),
+                None => CachePosition::None,
+            },
+        }
+    }
+
+    pub fn get_with_cost(&mut self, hash: &CryptoHash) -> (Option<Vec<u8>>, RetrievalCost) {
```

Comment:
I suggest that we create a new type for the return value here. It is a bit hard to read otherwise